src/HappyEyeBallsConnectionBuilder.php

```diff
@@ -0,0 +1,304 @@
+                    $uri .= '#' . $that->parts['fragment'];
+                }
+
+                $promise = $that->connector->connect($uri);
+                $promise->then($resolve, $reject);
+            },
+            function ($_, $reject) use (&$promise, &$resolved, $that) {
+                // cancellation should reject connection attempt
+                // (try to) cancel pending connection attempt
+                if ($resolved === null) {
+                    $reject(new \RuntimeException('Connection to ' . $that->uri . ' cancelled during connection attempt'));
+                }
+
+                if ($promise instanceof CancellablePromiseInterface) {
+                    // overwrite callback arguments for PHP7+ only, so they do not show
+                    // up in the Exception trace and do not cause a possible cyclic reference.
+                    $_ = $reject = null;
+
+                    $promise->cancel();
+                    $promise = null;
+                }
+            }
+        );
+    }
+
+    /**
+     * @internal
+     */
+    public function cleanUp()
+    {
+        /** @var CancellablePromiseInterface $promise */
+        foreach ($this->connectionPromises as $index => $connectionPromise) {
+            if ($connectionPromise instanceof CancellablePromiseInterface) {
+                $connectionPromise->cancel();
+            }
+            unset($this->connectionPromises[$index]);
+            $connectionPromise = null;
+        }
+        $this->connectionPromises = null;
+
+        /** @var CancellablePromiseInterface $promise */
+        foreach ($this->resolverPromises as $index => $resolverPromise) {
+            if ($resolverPromise instanceof CancellablePromiseInterface) {
+                $resolverPromise->cancel();
+            }
+            unset($this->resolverPromises[$index]);
+            $resolverPromise = null;
+        }
+        $this->resolverPromises = null;
+
+        if ($this->timer instanceof TimerInterface) {
+            $this->loop->cancelTimer($this->timer);
+            $this->timer = null;
+        }
+    }
+
+    /**
+     * @internal
+     */
+    public function hasBeenResolved()
+    {
+        foreach ($this->resolved as $typeHasBeenResolved) {
+            if ($typeHasBeenResolved === false) {
+                return false;
+
+                break;
+            }
+        }
+
+        return true;
+    }
+
+    /**
+     * @internal
+     */
+    public function pushIpsOntoQueue(array $ips)
+    {
+        $this->ipsCount += \count($ips);
+        $connectQueueStash = $this->connectQueue;
+        $this->connectQueue = array();
+        while (\count($connectQueueStash) > 0 || \count($ips) > 0) {
+            if (\count($ips) > 0) {
+                $this->connectQueue[] = \array_shift($ips);
+            }
+            if (\count($connectQueueStash) > 0) {
+                $this->connectQueue[] = \array_shift($connectQueueStash);
+            }
+        }
```

Comment:
I'm not sure I understand what I'm looking at tbqh.

Reply:
This mixes the new list of IP addresses (`$ips`) into the current one `$this->connectQueue`. This mixes the IPv4 addresses into the list of IPv6 addresses (or vise versa) in such a way that they alternate which IPvX is used to connect with.